# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: layout/base/nsPresContext.cpp
# Commit: e3f8526e0e2a
# Full Hash: e3f8526e0e2a08a4a8342dccedf05bebb3de1baf
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2021-09-08 21:39:05
# Regressor Bug: 1715501
# File Overlap Count: 1
# Description:
#   Bug 1715501 - patch 2 - Pass the presContext to platform font lookup methods so they can query it for font visibility. r=emilio
#   
#   This does not in itself change user-visible behavior, but is a foundation for bug
#   1715507 where we will make the behavior per-context instead of global. This is basically
#   just plumbing, passing a pointer to the presContext that's asking for fonts to be
# ==============================================================================

diff -r cfdd1c5d8d3f -r e3f8526e0e2a layout/base/nsPresContext.cpp
--- a/layout/base/nsPresContext.cpp	Wed Sep 08 09:35:16 2021 +0000
+++ b/layout/base/nsPresContext.cpp	Wed Sep 08 09:35:17 2021 +0000
@@ -270,6 +270,8 @@
     // fine to set here.
     mDynamicToolbarMaxHeight = StaticPrefs::layout_dynamic_toolbar_max_height();
   }
+
+  UpdateFontVisibility();
 }
 
 static const char* gExactCallbackPrefs[] = {
@@ -290,7 +292,8 @@
 
 static const char* gPrefixCallbackPrefs[] = {
     "font.", "browser.display.",    "browser.viewport.",
-    "bidi.", "gfx.font_rendering.", nullptr,
+    "bidi.", "gfx.font_rendering.", "layout.css.font-visibility.",
+    nullptr,
 };
 
 void nsPresContext::Destroy() {
@@ -638,6 +641,9 @@
   GetUserPreferences();
 
   FlushFontCache();
+  if (UpdateFontVisibility()) {
+    changeHint |= NS_STYLE_HINT_REFLOW;
+  }
 
   // Preferences require rerunning selector matching because we rebuild
   // the pref style sheet for some preference changes.
@@ -760,6 +766,19 @@
   return NS_OK;
 }
 
+bool nsPresContext::UpdateFontVisibility() {
+  FontVisibility oldValue = mFontVisibility;
+  if (StaticPrefs::privacy_resistFingerprinting()) {
+    mFontVisibility = FontVisibility::Base;
+  } else {
+    mFontVisibility = FontVisibility(
+        std::min(int32_t(FontVisibility::User),
+                 std::max(int32_t(FontVisibility::Base),
+                          StaticPrefs::layout_css_font_visibility_level())));
+  }
+  return mFontVisibility != oldValue;
+}
+
 void nsPresContext::InitFontCache() {
   if (!mFontCache) {
     mFontCache = new nsFontCache();